### ISSUE TITLE
errors under load with policy creation(ladon_action table)

### ISSUE THREAD

**Issue by alex-vento** [ory/keto#30]
Hello!
I made a load tests on policy creation and find some bugs with "duplicated pk error" on pg side.

After that I looked in the sql part of the source and find this:

```Go
QueryInsertPolicy: `INSERT INTO ladon_policy(id, description, effect, conditions) SELECT $1::varchar, $2, $3, $4 WHERE NOT EXISTS (SELECT 1 FROM ladon_policy WHERE id = $1);`,
```

So, it's bad to use ```GO WHERE NOT EXISTS``` it's need to replace on ```ON CONFLICT```

If you wanna to repeat it, just try to create many policies in N threads by send POST to "/policies" handler

**Comment by arekkas** [comment]
Did the policies you create have the same ID?

**Comment by arekkas** [comment]
If it were to be replace with `ON CONFLICT` it would be `ON CONFLICT DO NOTHING` because keys must be unique and we don't know if it's an error on your side (accidentally using the same key) or if you want to do an upsert. As we don't know, it's better to throw an error and let you know that the ID exists already.

**Comment by alex-vento** [comment]
This error not with policy, it's with action creation on policy create.

Like an example: You try to create N policies with action "read", on every policy you check is this action exists, and if not - create new action, so, in current realisation we have a bug with  `WHERE NOT EXIST` which returns on check, that action does not exist, but it's exist, if you will use ON CONFLIC DO NOTHING, this bug gones

**Comment by arekkas** [comment]
Oh I see, so the issue is that - due to rapid requests - there's a synchronization issue where `SELECT 1 FROM ladon_policy WHERE id = $1` yields `NULL` (or false) but once the query is actually executed, it's actually no longer false (because another transaction created the action). Did I get that right?

**Comment by alex-vento** [comment]
Yes, and this bug can be repeated, not only on new creation, it's repeated with exist action entries, because it's pg sync error, when you do queries with `WHERE NOT EXIST` you do not lock the table ;)

**Comment by arekkas** [comment]
Ok thanks for raising this issue, I'll take a closer look at this. I think `WHERE NOT EXIST` was used to make one query that works for both mysql and postgres, but apparently has synchronization issues.
